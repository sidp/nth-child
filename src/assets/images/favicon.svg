<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg">
	<style>
		.bar {
			fill: #edf2f7;
		}
		.bar.selected {
			fill: #37dc66;
		}
		@media (prefers-color-scheme: dark) {
			.bar {
				fill: #929fad;
			}
		}
	</style>
	<rect class="bar selected" x="1" y="1" width="14" height="2" rx="1"/>
	<rect class="bar" x="1" y="4" width="14" height="2" rx="1"/>
	<rect class="bar" x="1" y="7" width="14" height="2" rx="1"/>
	<rect class="bar selected" x="1" y="10" width="14" height="2" rx="1"/>
	<rect class="bar" x="1" y="13" width="14" height="2" rx="1"/>
</svg>
