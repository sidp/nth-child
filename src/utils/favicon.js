import { green, lightGray, gray } from './styles';

export default function favicon(pattern) {
	const matches = (i) => isMatched(pattern, i);

	return `data:image/svg+xml;utf8,<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg">
		<style>
			.selected {
				fill: ${svgColor(green)};
			}
			.inactive {
				fill: ${svgColor(lightGray)};
			}
			@media (prefers-color-scheme: dark) {
				.inactive {
					fill: ${svgColor(gray)};
				}
			}
		</style>
		<rect class="${matches(1) ? 'selected' : 'inactive'}" x="1" y="1" width="14" height="2" rx="1"/>
		<rect class="${matches(2) ? 'selected' : 'inactive'}" x="1" y="4" width="14" height="2" rx="1"/>
		<rect class="${matches(3) ? 'selected' : 'inactive'}" x="1" y="7" width="14" height="2" rx="1"/>
		<rect class="${matches(4) ? 'selected' : 'inactive'}" x="1" y="10" width="14" height="2" rx="1"/>
		<rect class="${matches(5) ? 'selected' : 'inactive'}" x="1" y="13" width="14" height="2" rx="1"/>
	</svg>`;
}

const isMatched = (pattern, i) => {
	pattern = pattern.trim();
	const seq = Array(50)
		.fill('')
		.map((_, i) => i);

	if (pattern === 'odd') {
		return i % 2 === 1;
	}

	if (pattern === 'even') {
		return i % 2 === 0;
	}

	if (pattern.indexOf('n') !== -1) {
		const parts = pattern.split('n');

		if (
			typeof parts[1] === 'string' &&
			parts[1].length > 0 &&
			!['+', '-'].includes(parts[1].charAt(0))
		) {
			return false;
		}

		const addition = parts[1] ? Number(parts[1]) : 0;

		return seq.some((s) => parts[0] * s + addition === i);
	}

	const no = Number(pattern);
	if (isNaN(no)) return false;
	return no === i;
};

const svgColor = (color) => encodeURIComponent(color);
