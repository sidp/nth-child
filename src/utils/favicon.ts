import { isMatched } from './pattern';
import { gray, green, lightGray } from './styles';

export function setFavicon(pattern: string) {
	const icoEl = document.querySelector<HTMLLinkElement>(
		'head link[rel="icon"][type="image/x-icon"]'
	);
	icoEl?.parentElement?.removeChild(icoEl);

	let iconEl = document.querySelector<HTMLLinkElement>(
		'head link[rel="icon"][type="image/svg+xml"]'
	);

	if (!iconEl) {
		iconEl = document.createElement('link');
		iconEl.setAttribute('rel', 'icon');
		iconEl.setAttribute('type', 'image/svg+xml');
		iconEl.setAttribute('href', favicon(pattern));
		document.head.appendChild(iconEl);
	} else {
		iconEl.setAttribute('href', favicon(pattern));
	}
}

export function favicon(pattern: string) {
	const matches = (i: number) => isMatched(pattern, i);

	return `data:image/svg+xml;utf8,<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg">
		<style>
			.bar {
				fill: ${svgColor(lightGray)};
			}
			.bar.selected {
				fill: ${svgColor(green)};
			}
			@media (prefers-color-scheme: dark) {
				.bar {
					fill: ${svgColor(gray)};
				}
			}
		</style>
		<rect class="bar ${
			matches(1) ? 'selected' : ''
		}" x="1" y="1" width="14" height="2" rx="1"/>
		<rect class="bar ${
			matches(2) ? 'selected' : ''
		}" x="1" y="4" width="14" height="2" rx="1"/>
		<rect class="bar ${
			matches(3) ? 'selected' : ''
		}" x="1" y="7" width="14" height="2" rx="1"/>
		<rect class="bar ${
			matches(4) ? 'selected' : ''
		}" x="1" y="10" width="14" height="2" rx="1"/>
		<rect class="bar ${
			matches(5) ? 'selected' : ''
		}" x="1" y="13" width="14" height="2" rx="1"/>
	</svg>`;
}

const svgColor = (color: string) => encodeURIComponent(color);
