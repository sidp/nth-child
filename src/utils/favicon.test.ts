import { describe, it, expect, beforeEach } from 'vitest';
import { favicon, setFavicon } from './favicon';
import { afterEach } from 'node:test';

describe('favicon', () => {
	it('should match previous snapshot', () => {
		expect(favicon('3n+1')).toMatchSnapshot();
		expect(favicon('6n-2')).toMatchSnapshot();
	});

	it('should return null for invalid pattern', () => {
		expect(favicon('hello')).toBeNull();
	});
});

describe('setFavicon', () => {
	beforeEach(() => {
		const iconElement = document.createElement('link');
		iconElement.rel = 'icon';
		iconElement.type = 'image/svg+xml';
		iconElement.href = '/favicon.svg';
		document.head.appendChild(iconElement);

		const icoEl = document.createElement('link');
		icoEl.rel = 'icon';
		icoEl.type = 'image/x-icon';
		document.head.appendChild(icoEl);
	});

	afterEach(() => {
		const iconElement = document.querySelector<HTMLLinkElement>(
			'head link[rel="icon"][type="image/svg+xml"]'
		);
		iconElement?.remove();
	});

	it('should set the favicon', () => {
		setFavicon('3n+1');
		const iconEl = document.querySelector<HTMLLinkElement>(
			'head link[rel="icon"][type="image/svg+xml"]'
		);
		expect(iconEl?.href).toMatch(/^data:image\/svg\+xml;utf8,<svg /);
	});

	it('should remove ico favicon', () => {
		setFavicon('3n+1');
		const icoElement = document.querySelector<HTMLLinkElement>(
			'head link[rel="icon"][type="image/x-icon"]'
		);
		expect(icoElement).toBeNull();
	});

	it('should create a link element if it does not exist', () => {
		document
			.querySelector<HTMLLinkElement>(
				'head link[rel="icon"][type="image/svg+xml"]'
			)
			?.remove();

		setFavicon('3n+1');

		expect(
			document.querySelector<HTMLLinkElement>(
				'head link[rel="icon"][type="image/svg+xml"]'
			)
		).not.toBeNull();
	});

	it('should not update favicon with invalid pattern', () => {
		const prevHref = document
			.querySelector<HTMLLinkElement>(
				'head link[rel="icon"][type="image/svg+xml"]'
			)
			?.getAttribute('href');

		setFavicon('hello');

		expect(
			document
				.querySelector<HTMLLinkElement>(
					'head link[rel="icon"][type="image/svg+xml"]'
				)
				?.getAttribute('href')
		).toEqual(prevHref);
	});
});
