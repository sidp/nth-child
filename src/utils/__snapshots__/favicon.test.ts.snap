// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`favicon > should match previous snapshot 1`] = `
"data:image/svg+xml;utf8,<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg">
		<style>
			.bar {
				fill: %23edf2f7;
			}
			.bar.selected {
				fill: %2337dc66;
			}
			@media (prefers-color-scheme: dark) {
				.bar {
					fill: %23929fad;
				}
			}
		</style>
		<rect class="bar selected" x="1" y="1" width="14" height="2" rx="1"/>
		<rect class="bar " x="1" y="4" width="14" height="2" rx="1"/>
		<rect class="bar " x="1" y="7" width="14" height="2" rx="1"/>
		<rect class="bar selected" x="1" y="10" width="14" height="2" rx="1"/>
		<rect class="bar " x="1" y="13" width="14" height="2" rx="1"/>
	</svg>"
`;

exports[`favicon > should match previous snapshot 2`] = `
"data:image/svg+xml;utf8,<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg">
		<style>
			.bar {
				fill: %23edf2f7;
			}
			.bar.selected {
				fill: %2337dc66;
			}
			@media (prefers-color-scheme: dark) {
				.bar {
					fill: %23929fad;
				}
			}
		</style>
		<rect class="bar " x="1" y="1" width="14" height="2" rx="1"/>
		<rect class="bar " x="1" y="4" width="14" height="2" rx="1"/>
		<rect class="bar " x="1" y="7" width="14" height="2" rx="1"/>
		<rect class="bar selected" x="1" y="10" width="14" height="2" rx="1"/>
		<rect class="bar " x="1" y="13" width="14" height="2" rx="1"/>
	</svg>"
`;
